File: smali/src/main/java/com/github/netomi/bat/smali/disassemble/InitializationDetector.java
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.smali.disassemble;

import com.github.netomi.bat.dexfile.*;
import com.github.netomi.bat.dexfile.instruction.DexInstruction;
import com.github.netomi.bat.dexfile.instruction.FieldInstruction;
import com.github.netomi.bat.dexfile.visitor.InstructionVisitor;

class      InitializationDetector
implements InstructionVisitor
{
    private final String name;
    private final String type;

    private boolean fieldIsSetInStaticInitializer;

    public InitializationDetector(String fieldName, String type) {
        this.name = fieldName;
        this.type = type;
    }

    public boolean fieldIsSetInStaticInitializer() {
        return fieldIsSetInStaticInitializer;
    }

    @Override
    public void visitAnyInstruction(DexFile dexFile, ClassDef classDef, EncodedMethod method, Code code, int offset, DexInstruction instruction) {}

    @Override
    public void visitFieldInstruction(DexFile dexFile, ClassDef classDef, EncodedMethod method, Code code, int offset, FieldInstruction instruction) {
        String mnemonic = instruction.getMnemonic();


        if (mnemonic.contains("sput")) {
            FieldID fieldID = instruction.getField(dexFile);

            if (fieldID.getName(dexFile).equals(name) &&
                fieldID.getType(dexFile).equals(type)) {
                fieldIsSetInStaticInitializer = true;
            }
        }

    }
}

File: smali/src/main/kotlin/com/github/netomi/bat/smali/disassemble/InitializationDetector.kt
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.smali.disassemble

import com.github.netomi.bat.dexfile.ClassDef
import com.github.netomi.bat.dexfile.Code
import com.github.netomi.bat.dexfile.DexFile
import com.github.netomi.bat.dexfile.EncodedMethod
import com.github.netomi.bat.dexfile.instruction.DexInstruction
import com.github.netomi.bat.dexfile.instruction.FieldInstruction
import com.github.netomi.bat.dexfile.instruction.visitor.InstructionVisitor

internal class InitializationDetector(private val name: String, private val type: String) : InstructionVisitor {
    var fieldIsSetInStaticInitializer = false
        private set

    override fun visitAnyInstruction(dexFile: DexFile, classDef: ClassDef, method: EncodedMethod, code: Code, offset: Int, instruction: DexInstruction) {}

    override fun visitFieldInstruction(dexFile: DexFile, classDef: ClassDef, method: EncodedMethod, code: Code, offset: Int, instruction: FieldInstruction) {
        val mnemonic = instruction.mnemonic

        if (mnemonic.contains("sput")) {
            val fieldID = instruction.getField(dexFile)
            if (fieldID.getName(dexFile) == name && fieldID.getType(dexFile) == type) {
                fieldIsSetInStaticInitializer = true
            }
        }
    }
}